Dynamic run for run-aqa GH action

j-braley opened the pull request:
Addresses #125.

Adds in functionality to run tests in parallel. This is accomplished by having a new function generate `parallelList.mk` and updating the `runAqaTest()` function run the suite of tests if the target includes `-f parallelList.mk`.

Any GitHub action which leverages this action handles distributing the `parallelList.mk` generated to child jobs (e.g., upload artifact/ download artifact).

Please let me know any questions, comments, or concerns 😄!

j-braley commented:
Just submitted the ECA.

karianna commented on src/runaqa.ts:
```diff
@@ -1,324 +1,388 @@
+import {ExecOptions} from '@actions/exec/lib/interfaces'
+
+let tempDirectory = process.env['RUNNER_TEMP'] || ''
+const IS_WINDOWS = process.platform === 'win32'
```
Is it always win32?  Can it be win64?

j-braley replied:
This is always win32. Its a Node.js functionality. 

From the docs:

```
Currently possible values are:

'aix'
'darwin'
'freebsd'
'linux'
'openbsd'
'sunos'
'win32'


https://nodejs.org/api/process.html#processplatform

karianna commented on src/runaqa.ts:
```diff
@@ -1,324 +1,388 @@
+  if (IS_WINDOWS) {
+    // On windows use the USERPROFILE env variable
+    baseLocation = process.env['USERPROFILE'] || 'C:\\'
+  } else if (process.platform === 'darwin') {
```
You've extracted the IS_WINDOWS, maybe extract IS_MAC

j-braley replied:
Will refactor

karianna commented on src/runaqa.ts:
```diff
@@ -1,324 +1,388 @@
+    aqasystemtestsRepo
+    );
+
+  const options: ExecOptions = {}
```
Add a comment here, what typically might get added?

j-braley replied:
Would defer to @sophia-guo to answer this.

karianna commented on src/runaqa.ts:
```diff
@@ -1,324 +1,388 @@
+    javaHome = process.env['JAVA_HOME'] as string
+  }
+  if (jdksource === 'install-jdk') {
+    // work with AdoptOpenJDK/install-sdk
```
Is that still AdoptOpenJDK or Adoptium?

j-braley replied:
Will update.

j-braley replied:
I believe this comment means that we are using the AdoptOpenJDK/install-jdk action. So I do not think there is anything to update here.

https://github.com/AdoptOpenJDK/install-jdk --> Action is still under AdoptOpenJDK/install-jdk

smlambert replied:
correct, install-jdk did not get transitioned over to Eclipse Foundation... also, at some point we may stop using it (as per https://github.com/adoptium/run-aqa/issues/122), but for now please leave as is

karianna commented on src/runaqa.ts:
```diff
@@ -1,324 +1,388 @@
+      `${tempDirectory}/ant-contrib/lib/ant-contrib.jar`,
+      `${process.env.ANT_HOME}\\lib`
+    )
+  } else if (process.platform === 'darwin') {
```
Would be good to have all of the O/S derivates - so it's clear that it's isMac, isSolaris, isLinux, isWindows and so forth

karianna commented on src/runaqa.ts:
```diff
@@ -1,324 +1,388 @@
+    }
+
+    if (fs.existsSync('/usr/bin/apt-get')) {
+      //disable apport
```
why?

smlambert replied:
If I change my Diff view to hide whitespace, I see a more accurate view of the code changes from @j-braley, I presume the editor being used has inserted additional or different whitespace, which causes it to look like the whole file changed.

j-braley replied:
That would make sense. I knew it was something I did 😆

smlambert replied:
notice how I blamed the IDE not you!  haha!

karianna reviewed (changes requested):
Functions could also do with more comments explaining roughly what each one does

j-braley commented on action.yml:
```diff
@@ -50,6 +50,12 @@ inputs:
+      description: 'Run jobs in parallel'
+      required: false
+  num_machines:
+      description: 'Number of machines to run parallel'
```
Will change.

j-braley commented on src/runaqa.ts:
```diff
@@ -1,324 +1,388 @@
+
+function getTestJdkHome(version: string, jdksource: string): string {
+  // Try JAVA_HOME first and then fall back to GITHUB actions default location
+  let javaHome = process.env[`JAVA_HOME_${version}_X64`] as string
```
I dont believe that GitHub has an aarch64 runner available. Will confirm.

j-braley replied:
I dont see aarch64 as an available runner at this time: [supported runners](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources)

j-braley replied:
Can confirm that currently **[only Linux Aarch64 is supported as a self-hosted runner](https://docs.github.com/en/actions/hosting-your-own-runners/about-self-hosted-runners#architectures)**.

j-braley commented on src/runaqa.ts:
```diff
@@ -1,324 +1,388 @@
+}
+
+function resetJDKHomeFromProperties(): void {
+  const pfile = `${process.env.GITHUB_WORKSPACE}/aqa-tests/job.properties`
```
Probably? I can make the refactor.

j-braley commented on src/runaqa.ts:
```diff
@@ -1,324 +1,388 @@
+  if (javaHome === undefined) {
+    core.error('JDK could not be found')
+  }
+  return javaHome
```
I do not believe so. Though the `core.error` should except and fail the build if `javaHome === undefined`

j-braley replied:
Set this to `core.setFailed('JDK could not be found')` - this should fail the build if the JDK cannot be found. 


(if this was not desired please let me know; I can revert)

j-braley commented on src/runaqa.ts:
```diff
@@ -1,324 +1,388 @@
+}
+
+// This function is an alternative of extra install step in workflow or alternative install action. This could also be implemented as github action
+async function installDependencyAndSetup(): Promise<void> {
```
I believe the "setup" part was moving ant-contrib around. Which is a dependency. I renamed the function to _hopefully_ something more descriptive.

j-braley commented:
I believe I have addressed most comments raised to the best of my knowledge.

Please let me know if there are any other changes required to get this across the finish line. 

Thanks!! 😄

karianna commented on src/runaqa.ts:
```diff
@@ -1,324 +1,475 @@
+        .toUpperCase()}_TARGET=${customTarget}`
+      await exec.exec('make', [`${target}`, `${customOption}`], options)
+    }
+    else if (target.includes('-f parallelList.mk')){
```
```suggestion
    else if (target.includes('-f parallelList.mk')) {

```

j-braley replied:
Refactored.

karianna commented on src/runaqa.ts:
```diff
@@ -1,324 +1,475 @@
+    }
+    else if (target.includes('-f parallelList.mk')){
+      // move the parallelList to TKG/
+      if (IS_WINDOWS){
```
```suggestion
      if (IS_WINDOWS) {

```

karianna commented on src/runaqa.ts:
```diff
@@ -1,324 +1,475 @@
+      await exec.exec('make', [`${target}`], options)
+    }
+  } catch (error) {
+    if (error instanceof Error){
```
```suggestion
    if (error instanceof Error) {

```

j-braley replied:
Refactored.

karianna commented on src/runaqa.ts:
```diff
@@ -1,324 +1,475 @@
+    if (error instanceof Error){
+      core.setFailed(error.message)
+    }
+    else{
```
```suggestion
    else {

```

j-braley replied:
Refactored.

karianna commented on src/runaqa.ts:
```diff
@@ -1,324 +1,475 @@
+        core.addPath(`C:\\cygwin64\\bin`)
+      }
+    } catch (error) {
+      if (error instanceof Error){
```
```suggestion
      if (error instanceof Error) {

```

j-braley replied:
Refactored.

karianna commented on src/runaqa.ts:
```diff
@@ -1,324 +1,475 @@
+      if (error instanceof Error){
+        core.warning(error.message)
+      }
+      else{
```
```suggestion
      else {

```

karianna commented on src/runaqa.ts:
```diff
@@ -1,324 +1,475 @@
+    `git clone --depth 1 -b ${repoBranch[1]} https://github.com/${repoBranch[0]}.git`
+  )
+  process.chdir('aqa-tests')
+  // workaround until TKG can download the artifacts required for Windows
```
Do we have an issue number to refer to here?

j-braley replied:
Negative. Will open one!

j-braley replied:
Opened: https://github.com/adoptium/TKG/issues/342

karianna commented on src/runaqa.ts:
```diff
@@ -1,324 +1,475 @@
+  )
+  process.chdir('aqa-tests')
+  // workaround until TKG can download the artifacts required for Windows
+  if (IS_WINDOWS && buildList != ''){
```
```suggestion
  if (IS_WINDOWS && buildList != '') {

```

j-braley replied:
Refactored.

karianna commented on src/runaqa.ts:
```diff
@@ -1,324 +1,475 @@
+  process.chdir('aqa-tests')
+  // workaround until TKG can download the artifacts required for Windows
+  if (IS_WINDOWS && buildList != ''){
+    if (buildList == 'system'){
```
```suggestion
    if (buildList == 'system') {

```

j-braley replied:
Refactored.

karianna commented on src/runaqa.ts:
```diff
@@ -1,324 +1,475 @@
+      await exec.exec(`git clone -q https://github.com/adoptium/STF.git`) // points to master/main
+      process.chdir('../')
+    }
+    if (buildList == 'openjdk' && version != ''){
```
```suggestion
    if (buildList == 'openjdk' && version != '') {

```

karianna replied:
Also should that check be === ?

j-braley replied:
Refactored.

karianna commented on src/runaqa.ts:
```diff
@@ -1,324 +1,475 @@
+    }
+    if (buildList == 'openjdk' && version != ''){
+      process.chdir('openjdk')
+      await exec.exec(`git clone --depth 1 -q --reference-if-able ${process.env.GITHUB_WORKSPACE}/openjdk_cache https://github.com/adoptium/jdk${version}.git openjdk-jdk`)
```
Do you want the `openjdk-jdk` target to have the version info appended as well?

karianna commented on src/runaqa.ts:
```diff
@@ -1,324 +1,475 @@
+  tkgRepo: string,
+  vendorTestParams: string,
+  aqasystemtestsRepo: string
+  ):  Promise<void>{
```
```suggestion
  ):  Promise<void> {

```

j-braley replied:
Refactored.

karianna commented on src/runaqa.ts:
```diff
@@ -1,324 +1,475 @@
+    await installPlatformDependencies();
+    setSpec();
+    process.env.BUILD_LIST = buildList;
+    if ((jdksource === 'upstream' ||
```
Could extract this into a small private method?

j-braley replied:
Refactored.

karianna commented on src/runaqa.ts:
```diff
@@ -1,324 +1,475 @@
+    await runGetSh(tkgRepo, openj9Repo, vendorTestParams, jdksource, customizedSdkUrl, sdkdir);
+    resetJDKHomeFromProperties();
+
+    //Get Dependencies, using /*zip*/dependents.zip to avoid loop every available files
```
```suggestion
    // Get Dependencies, using /*zip*/dependents.zip to avoid loop every available files

```

j-braley replied:
Refactored.

karianna commented on src/aqa.ts:
```diff
@@ -73,23 +74,46 @@ async function run(): Promise<void> {
-    )
+    if(runParallel === 'true'){
+
+      await runaqa.setupParallelEnv(
```
Does setupParallelEnv eventually call runaqaTest?  If so, is there a reason we didn't add an extra runInParallel boolean argument to runaqaTest instead?

j-braley replied:
It does not. The way we achieve parallelism in GitHub is to have one runner generate the `parallelList.mk` for the specific suite of tests cases. This file is uploaded to the workflow as an artifact. After which we spawn several other machines up which download the `parallelList.mk` with a specific list (e.g., testList_0) for which they are to run their tests. 

See here: https://github.com/j-braley/aqafer/actions/runs/2793593197 under "artifacts". 

The `aqafer` GH action contains the logic for downloading/running the test cases in parallel using a `strategy: matrix`

karianna commented on src/runaqa.ts:
```diff
@@ -1,324 +1,475 @@
+}
+
+/**
+ * This function is an alternative of extra install step in workflow or alternative install action. This could also be implemented as github action
```
I's add an issue in the run-aqa repo for this (create extraGH Action)

j-braley replied:
Will do

karianna commented on src/runaqa.ts:
```diff
@@ -1,324 +1,475 @@
+  if (IS_WINDOWS && buildList != ''){
+    if (buildList == 'system'){
+      process.chdir('system')
+      await exec.exec(`git clone -q https://github.com/adoptium/aqa-systemtest.git`)  // points to master/main
```
master or main, which one is used now?

j-braley replied:
master.

Fixed.

karianna commented on src/runaqa.ts:
```diff
@@ -1,324 +1,475 @@
+    }
+    if (buildList == 'openjdk' && version != ''){
+      process.chdir('openjdk')
+      await exec.exec(`git clone --depth 1 -q --reference-if-able ${process.env.GITHUB_WORKSPACE}/openjdk_cache https://github.com/adoptium/jdk${version}.git openjdk-jdk`)
```
Worth adding a comment as to what that git command does (flags included)

j-braley replied:
Done!

karianna commented on src/runaqa.ts:
```diff
@@ -1,324 +1,475 @@
+    resetJDKHomeFromProperties();
+
+    //Get Dependencies, using /*zip*/dependents.zip to avoid loop every available files
+    let dependents = await tc.downloadTool('https://ci.adoptopenjdk.net/view/all/job/test.getDependency/lastSuccessfulBuild/artifact//*zip*/dependents.zip');
```
should we parametrize the ci.adoptopenjdk.net URL so it can be overridden?

j-braley replied:
which part of this could be parameterized? Might we have to add additional args to the GH action which calls this action? Or some additional variables in the function?

karianna replied:
I'm just wondering if you were to run this in house with no access to ci.adoptoipenjdk.net but an internal clone instead, does that help?  Probably something for a future PR though :-)

smlambert replied:
I believe we put the behaviour to be if not able to access the public server, then download each of the dependencies individually.  Again, those would need access to public URLs.

karianna replied:
OK, we can resolve this later if required (we'll be trying in house over the coming weeks and can adjust)

karianna commented on src/runaqa.ts:
```diff
@@ -1,324 +1,475 @@
+
+    //Get Dependencies, using /*zip*/dependents.zip to avoid loop every available files
+    let dependents = await tc.downloadTool('https://ci.adoptopenjdk.net/view/all/job/test.getDependency/lastSuccessfulBuild/artifact//*zip*/dependents.zip');
+    let sevenzexe = '7z';
```
This seems to be a linux specific set up only.  Is that the only support O/S for the AQAVit runner?

j-braley replied:
We have support for Windows/MacOS/Linux (non-ARM).

j-braley commented on src/aqa.ts:
```diff
@@ -73,23 +74,46 @@ async function run(): Promise<void> {
+    if (error instanceof Error){
+      core.setFailed(error.message)
+    }
+    else{
```
Refactored.

karianna commented:
@j-braley Looks like more failing GH Actions though.

j-braley commented:
Updated GH action to run `_DaaLoadTest_daa1_5m` instead of `_DaaLoadTest_daa1` for sys tests + bumped up windows version. Changes based on changes found in https://github.com/adoptium/run-aqa/pull/131.

j-braley commented:
@karianna or @smlambert can I get an approval to run the extra GH actions? 

I made changes to the underlying action to run `_DaaLoadTest_daa1_5m` instead of `_DaaLoadTest_daa1` based on https://github.com/adoptium/run-aqa/pull/131. Also bumped up the windows version (also found in the linked PR).

renfeiw commented on src/aqa.ts:
```diff
@@ -1,95 +1,119 @@
+        tkgRepo,
+        vendorTestParams,
+        aqasystemtestsRepo,
+        numMachines
```
FYI, for parallel, we can also use the testTime parameter instead of the numMachines. But I think this doesn't have to be in this pr, can be a future improvement.

j-braley replied:
Ok, cool.

renfeiw commented on src/runaqa.ts:
```diff
@@ -312,13 +359,129 @@ async function runGetSh(
+
+  await setupTestEnv(version, jdksource, customizedSdkUrl, sdkdir, buildList, aqatestsRepo, openj9Repo, tkgRepo, vendorTestParams, aqasystemtestsRepo);
+  process.chdir('TKG');
+  process.env.PARALLEL_OPTIONS = `PARALLEL_OPTIONS=TEST=${buildList} TEST_TIME= NUM_MACHINES=${numMachines}`;
```
TEST should equal to the test target.

j-braley replied:
I believe `BUILD_LIST` here would be the target? Since the values of `BUILD_LIST` would be one of `openjdk, functional, system, perf`. Or am I mistaken? If the verbiage does not make sense that could be adjusted. However, I believe that this value matches what is set in the env variable `BUILD_LIST`.

renfeiw replied:
`BUILD_LIST` is not the target, it is the directory to be compiled. Some target and `BUILD_LIST` happen to share the same name, for example `functional`, but target also could be `sanity.functional` which is different from BUILD_LIST

j-braley replied:
Thank you for the explanation. I will update the code.

sophia-guo commented:
Enable github action runs and wait for the results

sophia-guo commented on src/aqa.ts:
```diff
@@ -73,23 +74,47 @@ async function run(): Promise<void> {
-      vendorTestParams,
-      aqasystemtestsRepo
-    )
+    if(runParallel === 'true') {
```
Could we enhance this condition to check the numMachines in case that runParalllel is set true and numMachines is set to 1?

sophia-guo commented on src/runaqa.ts:
```diff
@@ -8,21 +8,36 @@ import {ExecOptions} from '@actions/exec/lib/interfaces'
+ * @param  {string} jdksource Source for JDK
+ * @param  {[string]} customizedSdkUrl Download link for JDK binaries
+ * @param  {[string]} sdkdir Directory for SDK
+ * @param  {[string]} buildList AQAvit Test suite
```
Could we also add @param target: string, customTarget: string ?

sophia-guo commented on src/runaqa.ts:
```diff
@@ -110,27 +76,52 @@ export async function runaqaTest(
+    }
+    else if (target.includes('-f parallelList.mk')) {
+      // move the parallelList to TKG/
+      if (IS_WINDOWS) {
```
As all test environment setup is moving to setupTestEnv() I would suggest moving this logic to setupTestEnv().

sophia-guo commented on action.yml:
```diff
@@ -1,55 +1,61 @@
+      description: 'Comma-separated list of vendor SHAs'
+      required: false
+  run_parallel:
+      description: 'Run jobs in parallel'
```
This parameter is a little bit confusing. Looking at [the workflow example ](https://github.com/j-braley/aqafer/actions/runs/2826544204/workflow) the parameter is used to generate file parallelList.mk not `run jobs in parallel`. I'm thinking maybe we should rename it to make a more clearer?  `run jobs in parallel` is implemented by github action matrix.

sophia-guo commented:
To make it easy to use/understand the example workflow can also be committed. As action.yml itself can not clearly demo how the feature is used.

j-braley commented:
Apparently, I am learning that GitHub has limits on reviewers 😦

karianna commented:
@j-braley you'll need to rebase again sorry!

sophia-guo reviewed (approved):
LGTM.
